File: examples/04-kafka/go/kafka_consumer_activity.go
package kafka

import (
	"context"
	"fmt"

	"github.com/segmentio/kafka-go"
	"go.temporal.io/sdk/activity"
)

// A polling consumer lives inside an Activity (blocking I/O, never Workflow
// code). We use ReadMessage (not auto-commit) and CommitMessages only after the
// Activity's work succeeds — giving at-least-once delivery into Temporal.
func PollBatch(ctx context.Context, topic string) ([]string, error) {
	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:     []string{"localhost:9092"},
		Topic:       topic,
		GroupID:     "temporal-poll-batch",
		StartOffset: kafka.FirstOffset,
	})
	defer reader.Close()

	var values []string
	var batch []kafka.Message
	for len(values) < 100 {
		msg, err := reader.FetchMessage(ctx)
		if err != nil {
			return nil, err
		}
		values = append(values, string(msg.Value))
		batch = append(batch, msg)
		// Heartbeat the offset so a retry resumes from here.
		activity.RecordHeartbeat(ctx, fmt.Sprintf("%s:%d:%d", msg.Topic, msg.Partition, msg.Offset))
	}

	// Commit only after Activity work succeeds.
	if err := reader.CommitMessages(ctx, batch...); err != nil {
		return nil, err
	}
	return values, nil
}

File: examples/04-kafka/python/kafka_consumer_activity.py
from temporalio import activity

# A polling consumer lives inside an Activity (it is blocking I/O, never Workflow
# code). kafka-python's KafkaConsumer is created with enable_auto_commit=False so
# we commit only after the Activity's work succeeds — at-least-once into Temporal.


@activity.defn
async def poll_batch(topic: str) -> list[str]:
    from kafka import KafkaConsumer

    consumer = KafkaConsumer(
        topic,
        bootstrap_servers="localhost:9092",
        enable_auto_commit=False,
        auto_offset_reset="earliest",
        value_deserializer=lambda b: b.decode(),
    )
    values: list[str] = []
    try:
        while len(values) < 100:
            for tp, records in consumer.poll(timeout_ms=1000).items():
                for record in records:
                    values.append(record.value)
                    # Heartbeat the offset so a retry resumes from here.
                    activity.heartbeat(f"{record.topic}:{record.partition}:{record.offset}")
        # Commit only after Activity work succeeds.
        consumer.commit()
        return values
    finally:
        consumer.close()
